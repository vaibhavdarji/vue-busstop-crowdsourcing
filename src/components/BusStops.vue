<template>
  <v-card :id="Constants.APP_TITLE.BS_LIST_ID">
    <v-container grid-list-xl fluid>
      <v-layout v-if="isLoading" row wrap justify-center align-center>
        <v-progress-circular indeterminate color="primary"></v-progress-circular>
      </v-layout>
      <template v-if="errorMsg">
        <custom-error :msg="errorMsg" @onErrorBtnClick="onRefresh"></custom-error>
      </template>

      <v-layout row wrap v-else>
        <v-flex xs4 v-for="stop in totalStops" :key="stop.stopId" :class="Constants.APP_BS_ITEMS.BS_CARD_CLS">
          <bus-stop :stop="stop"></bus-stop>
        </v-flex>
      </v-layout>
    </v-container>
  </v-card>
</template>

<script>
import { mapGetters, mapActions } from "vuex";
import Constants from '@/constants';
import CustomButton from "@/components/Button.vue";
import BusStop from "@/components/BusStop.vue";
import CustomError from "@/components/Error.vue";
import {
  GET_ALL_STOPS
} from '@/actions/types';
export default {
  name: "BusStops",
  data() {
    return {
      errorMsg: "",
      doRefresh: false
    };
  },

  components: {
    BusStop,
    CustomButton,
    CustomError
  },

  asyncComputed: {
    totalStops: {
      get() {
        return this.busStops.length
          ? this.busStops
          : this.getAllStops()
              .then(data => {
                this.doRefresh = false;
                return data;
              })
              .catch(error => {
                this.errorMsg = error.message;
                this.doRefresh = false;
              });
      },

      default() {
        return [];
      },
      watch() {
        this.doRefresh
      }
    }
  },
  computed: {
    ...mapGetters(["isLoading", "busStops"]),
    Constants () {
      return Constants;
    }
  },

  methods: {
    ...mapActions({
      "getAllStops": GET_ALL_STOPS
    }),

    onRefresh() {
      this.errorMsg = "";
      this.doRefresh = true;
    }
  }
};
</script>

<style>
.v-progress-linear {
  border-radius: 10px;
}
</style>


