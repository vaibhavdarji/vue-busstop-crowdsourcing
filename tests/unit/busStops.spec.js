import { shallowMount, mount } from "@vue/test-utils";
import BusStops from '@/components/BusStops.vue';
import Vue from "vue";
import Vuex from "vuex";
import Vuetify from "vuetify";

Vue.use(Vuex);
Vue.use(Vuetify);

const dummyData = [
  { stopId: 1, lat: 33.760262, lng: -84.384706, donationsRaisedInDollars: 0, name: 'Hertz at Portman Blvd' },
  { stopId: 2, lat: 33.760138, lng: -84.388043, donationsRaisedInDollars: 0, name: 'Peachtree Center Mall' }
];

describe("BusStops.vue", () => {

  let  state, getters, actions, mutations, store;
  beforeEach(() => {
    state = {
      isLoading: false,
      busStops: []
    };

    mutations = {
      setLoading (loading) {
        state.isLoading = loading;
      },

      setStops (stops) {
        state.busStops = stops;
      }
    };

    actions = {
      getAllStops () {
        mutations.setLoading(true);
        mutations.setStops(dummyData);
        mutations.setLoading(false);
      }
    };

    getters = {
      isLoading: state => state.isLoading,
      busStops: state => state.busStops
    };

    store = new Vuex.Store({
      state,
      actions,
      getters
    });
  });

  it("should match snapshot", () => {
    const wrapper = shallowMount(BusStops, {
      store,
      Vue
    });
    expect(wrapper.element).toMatchSnapshot();
  });

  it("should load data", () => {
    actions.getAllStops();

    expect(state.isLoading).toBe(false);
    expect(state.busStops.length).toBe(dummyData.length);
  });

  it("should match data with no of DOM card elements generated", () => {
    actions.getAllStops();
    const wrapper = shallowMount(BusStops, {
      store,
      Vue
    });

    expect(wrapper.vm.busStops.length).toBe(dummyData.length);
    
  });
});