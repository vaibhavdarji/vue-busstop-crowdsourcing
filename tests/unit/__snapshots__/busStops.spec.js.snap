// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BusStops.vue should match snapshot 1`] = `
<v-card-stub
  id="bsc-app-bs-list"
  tag="div"
>
  <v-container-stub
    fluid=""
    grid-list-xl=""
    tag="div"
  >
    <!---->
     
    <v-layout-stub
      row=""
      tag="div"
      wrap=""
    />
  </v-container-stub>
</v-card-stub>
`;
